openssl header not found for dune build

I have encounter this issue on my mac that when I tried to install piaf, during dune build, it failed for such error that openssl's `rand.h` and `sha.h` not found, even though openssl@3 is installed via homebrew and following path is present `/opt/homebrew/Cellar/openssl@3/3.4.0/include/openssl/` for header file reference and I ran brew `brew link --force openssl@3` to make sure it's linked, but still failed. 

computer: apple macbook air m2
OS: macos Sequoia 15.1.1
package manager: brew
shell: bash

A temporary fix is put `export CPATH="/opt/homebrew/Cellar/openssl@3/3.4.0/include"` in `~/.bash_profile`

I have pkgconf, and facing the same issue:

```
$ brew info pkgconf openssl
==> pkgconf: stable 2.3.0 (bottled), HEAD
...
Installed
...
==> openssl@3: stable 3.4.0 (bottled)
...
Installed
```

Error:

```
$ opam install --yes piaf              
The following actions will be performed:
  ∗ install piaf 0.2.0

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><>  🐫 
⬇ retrieved piaf.0.2.0  (cached)
[ERROR] The compilation of piaf.0.2.0 failed at "dune build -p piaf -j 7 @install".

#=== ERROR while compiling piaf.0.2.0 =========================================#
# context     2.1.6 | macos/arm64 | ocaml-base-compiler.5.2.1 | https://opam.ocaml.org#a66a58bdc3f82c934381148025d7ae4282c5ef1f
# path        ~/.opam/dream-dev/.opam-switch/build/piaf.0.2.0
# command     ~/.opam/opam-init/hooks/sandbox.sh build dune build -p piaf -j 7 @install
# exit-code   1
# env-file    ~/.opam/log/piaf-74964-d7a2ce.env
# output-file ~/.opam/log/piaf-74964-d7a2ce.out
### output ###
# [...]
# lib/stubs/piaf_openssl_sha.c:37:10: fatal error: 'openssl/sha.h' file not found
# #include <openssl/sha.h>
#          ^~~~~~~~~~~~~~~
# 1 error generated.
# File "lib/dune", line 24, characters 28-45:
# 24 |   (names piaf_bigarray_read piaf_openssl_rand piaf_openssl_sha piaf_fdutils)))
#                                  ^^^^^^^^^^^^^^^^^
# (cd _build/default/lib && /usr/bin/cc -O2 -fno-strict-aliasing -fwrapv -pthread -pthread -D_FILE_OFFSET_BITS=64 -O2 -fno-strict-aliasing -fwrapv -pthread -pthread -g -I /Users/yawarquadiramin/.opam/dream-dev/lib/ocaml -I /Users/yawarquadiramin/.opam/dream-dev/lib/angstrom -I /Users/yawarquadiramin/.opam/dream-dev/lib/base64 -I /Users/yawarquadiramin/.opam/dream-dev/lib/base64/rfc2045 -I /User[...]
# lib/stubs/piaf_openssl_rand.c:37:10: fatal error: 'openssl/rand.h' file not found
# #include <openssl/rand.h>
#          ^~~~~~~~~~~~~~~~
# 1 error generated.
```

--- a/dune-project
+++ b/dune-project
@@ -1,12 +1,10 @@
-(lang dune 2.8)
+(lang dune 3.0)
 
 (using dune_site 0.1)
 
 (name piaf)
 
 (generate_opam_files true)
-
-(use_standard_c_and_cxx_flags true)
 
 (source
  (github anmonteiro/piaf))

--- a/dune-project
+++ b/dune-project
@@ -27,6 +27,7 @@
  (depends
   (ocaml
    (>= "5.1"))
+  dune-configurator
   logs
   (eio-ssl
    (>= 0.3.0))

--- a/dune-project
+++ b/dune-project
@@ -6,6 +6,8 @@
 
 (generate_opam_files true)
 
+(use_standard_c_and_cxx_flags true)
+
 (source
  (github anmonteiro/piaf))
 

--- a/lib/cflags/cflags.ml
+++ b/lib/cflags/cflags.ml
@@ -1,0 +1,29 @@
+module C = Configurator.V1
+
+let directory_exists fsp = Sys.file_exists fsp && Sys.is_directory fsp
+
+let default c : C.Pkg_config.package_conf =
+  if C.ocaml_config_var_exn c "system" = "macosx"
+  then
+    if directory_exists "/usr/local/opt/openssl"
+    then
+      { libs = [ "-L/usr/local/opt/openssl/lib" ]
+      ; cflags = [ "-I/usr/local/opt/openssl/include" ]
+      }
+    else { libs = [ "-L/opt/local/lib" ]; cflags = [ "-I/opt/local/include" ] }
+  else { libs = [ "-lssl"; "-lcrypto" ]; cflags = [] }
+
+let () =
+  C.main ~name:"ssl" (fun c ->
+      let default = default c in
+      let conf =
+        match C.Pkg_config.get c with
+        | None -> default
+        | Some pc ->
+          (match C.Pkg_config.query pc ~package:"openssl" with
+          | Some s -> s
+          | None -> default)
+      in
+      C.Flags.write_sexp "c_library_flags.sexp" conf.libs;
+      C.Flags.write_sexp "c_flags.sexp" conf.cflags)
+

--- a/lib/cflags/cflags.ml
+++ b/lib/cflags/cflags.ml
@@ -15,15 +15,14 @@ let default c : C.Pkg_config.package_conf =
 
 let () =
   C.main ~name:"ssl" (fun c ->
-      let default = default c in
-      let conf =
-        match C.Pkg_config.get c with
-        | None -> default
-        | Some pc ->
-          (match C.Pkg_config.query pc ~package:"openssl" with
-          | Some s -> s
-          | None -> default)
-      in
-      C.Flags.write_sexp "c_library_flags.sexp" conf.libs;
-      C.Flags.write_sexp "c_flags.sexp" conf.cflags)
-
+    let default = default c in
+    let conf =
+      match C.Pkg_config.get c with
+      | None -> default
+      | Some pc ->
+        (match C.Pkg_config.query pc ~package:"openssl" with
+        | Some s -> s
+        | None -> default)
+    in
+    C.Flags.write_sexp "c_library_flags.sexp" conf.libs;
+    C.Flags.write_sexp "c_flags.sexp" conf.cflags)

--- a/lib/cflags/dune
+++ b/lib/cflags/dune
@@ -1,0 +1,4 @@
+(executable
+ (name cflags)
+ (libraries dune-configurator))
+

--- a/lib/cflags/dune
+++ b/lib/cflags/dune
@@ -1,4 +1,3 @@
 (executable
  (name cflags)
  (libraries dune-configurator))
-

--- a/lib/dune
+++ b/lib/dune
@@ -21,4 +21,13 @@
   piaf.stream)
  (foreign_stubs
   (language c)
-  (names piaf_bigarray_read piaf_openssl_rand piaf_openssl_sha piaf_fdutils)))
+  (names piaf_bigarray_read piaf_openssl_rand piaf_openssl_sha piaf_fdutils)
+  (flags (:include c_flags.sexp)))
+ (c_library_flags
+  (:include c_library_flags.sexp)))
+
+(rule
+ (targets c_flags.sexp c_library_flags.sexp)
+ (action
+  (run ./cflags/cflags.exe)))
+

--- a/lib/dune
+++ b/lib/dune
@@ -22,7 +22,8 @@
  (foreign_stubs
   (language c)
   (names piaf_bigarray_read piaf_openssl_rand piaf_openssl_sha piaf_fdutils)
-  (flags (:include c_flags.sexp)))
+  (flags
+   (:include c_flags.sexp)))
  (c_library_flags
   (:include c_library_flags.sexp)))
 
@@ -30,4 +31,3 @@
  (targets c_flags.sexp c_library_flags.sexp)
  (action
   (run ./cflags/cflags.exe)))
-

--- a/nix/ci/test.nix
+++ b/nix/ci/test.nix
@@ -94,6 +94,7 @@ let
         dune
         findlib
         ocamlformat
+        dune-configurator
       ]);
       doCheck = true;
       checkPhase = ''

--- a/nix/default.nix
+++ b/nix/default.nix
@@ -22,6 +22,8 @@ rec {
       ];
     };
 
+    buildInputs = with ocamlPackages; [ dune-configurator ];
+
     propagatedBuildInputs = with ocamlPackages; [
       logs
       eio-ssl

--- a/piaf.opam
+++ b/piaf.opam
@@ -9,7 +9,7 @@ license: "BSD-3-clause"
 homepage: "https://github.com/anmonteiro/piaf"
 bug-reports: "https://github.com/anmonteiro/piaf/issues"
 depends: [
-  "dune" {>= "2.8"}
+  "dune" {>= "3.0"}
   "ocaml" {>= "5.1"}
   "logs"
   "eio-ssl" {>= "0.3.0"}
@@ -38,9 +38,11 @@ build: [
     name
     "-j"
     jobs
+    "--promote-install-files=false"
     "@install"
     "@runtest" {with-test}
     "@doc" {with-doc}
   ]
+  ["dune" "install" "-p" name "--create-install-files" name]
 ]
 dev-repo: "git+https://github.com/anmonteiro/piaf.git"

--- a/piaf.opam
+++ b/piaf.opam
@@ -11,6 +11,7 @@ bug-reports: "https://github.com/anmonteiro/piaf/issues"
 depends: [
   "dune" {>= "3.0"}
   "ocaml" {>= "5.1"}
+  "dune-configurator"
   "logs"
   "eio-ssl" {>= "0.3.0"}
   "magic-mime"

